File: Scala/test/appendix/LazyEvaluationTests.scala
package appendix

import mocks.Appendix.computed
import org.scalactic.Tolerance
import org.scalatest.funsuite.AnyFunSuite
import tinyscalautils.assertions.in

import java.math.BigInteger
import java.util as ju
import scala.annotation.tailrec
import scala.jdk.CollectionConverters.IterableHasAsScala
import scala.jdk.OptionConverters.RichOptional

class LazyEvaluationTests extends AnyFunSuite with Tolerance:
   import LazyEvaluation.*
   import LazyEvaluationJ.*

   test("fill/tabulate (Scala)") {
      assert(LazyEvaluation.demo1 == List("X", "XX", "XXX", "XXXX", "XXXXX"))
      @tailrec def loop(): Boolean = LazyEvaluation.demo2.toSet.size != 1 || loop()
      assert(loop())
   }

   test("fill/tabulate (Java)") {
      assert(LazyEvaluationJ.demo1 == ju.List.of("X", "XX", "XXX", "XXXX", "XXXXX"))
      @tailrec def loop(): Boolean = LazyEvaluationJ.demo2.asScala.toSet.size != 1 || loop()
      assert(loop())
   }

   test("lazy val") {
      computed = 0
      assert(variable == 1)
      assert(computed == 1)
      assert(variable == 1)
      assert(computed == 1)
   }

   test("timeOf") {
      assert(time === 1.0 +- 0.1)
   }

   test("orElse") {
      assert(str in Set("a", "42", "B"))
      assert(optStr.toScala.forall(str => str in Set("a", "42", "B")))
   }

   test("collatz") {
      assert(collatz(BigInteger.valueOf(27)) == 111)
      val n = BigInteger("992774507105260663893249807781832616822016143650134730933270")
      assert(collatz(n) == 2632)
   }

File: Kotlin/test/appendix/LazyEvaluationTests.kt
package appendix

import appendix.LazyEvaluation.collatz
import appendix.LazyEvaluation.demo1
import appendix.LazyEvaluation.demo2
import appendix.LazyEvaluation.demo3
import appendix.LazyEvaluation.time
import appendix.LazyEvaluation.variable
import mocks.computed
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class LazyEvaluationTests {
  @Test
  fun testList() {
    assertEquals(listOf("X", "XX", "XXX", "XXXX", "XXXXX"), demo1())
    fun loop(demo: () -> List<Int>): Boolean = demo().toSet().size != 1 || loop(demo)
    assertTrue(loop(::demo2))
    assertTrue(loop(::demo3))
  }

  @Test
  fun testTimeOf() {
    assertEquals(1.0, time, 0.1)
  }

  @Test
  fun testCollatz() {
    assertEquals(111, collatz(BigInteger("27")))
    val n = BigInteger("992774507105260663893249807781832616822016143650134730933270")
    assertEquals(2632, collatz(n))
  }

  @Test
  fun testLazy() {
    computed = 0
    assertEquals(1, variable)
    assertEquals(1, computed)
    assertEquals(1, variable)
    assertEquals(1, computed)
  }
}
